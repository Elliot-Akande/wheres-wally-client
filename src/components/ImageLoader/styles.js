import { styled } from "styled-components";

export const Placeholder = styled.div`
  width: 100%;
  height: 100%;

  display: flex;
  align-items: center;
  justify-content: center;

  ${(props) => props.$customStyles}
`;

export const Image = styled.img`
  transition: 0.3s;
  opacity: ${(props) => (props.$loading ? 0 : 1)};

  ${(props) => props.$customStyles};
`;
