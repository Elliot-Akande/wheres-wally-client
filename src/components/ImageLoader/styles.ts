import { styled, Interpolation } from "styled-components";

export const Placeholder = styled.div<{
  $customStyles?: Interpolation<React.CSSProperties>,
}>`
  width: 100%;
  height: 100%;

  display: flex;
  align-items: center;
  justify-content: center;

  ${({$customStyles}) => $customStyles}
`;

export const Image = styled.img<{
  $customStyles?: Interpolation<React.CSSProperties>,
  $loading: boolean,
}>`
  transition: 0.3s;
  opacity: ${({$loading}) => ($loading ? 0 : 1)};

  ${({$customStyles}) => $customStyles};
`;
