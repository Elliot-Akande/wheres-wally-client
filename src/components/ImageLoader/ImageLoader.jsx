import PropTypes from "prop-types";
import { useState } from "react";
import { ScaleLoader } from "react-spinners";
import { Image, Placeholder } from "./styles";

const ImageLoader = ({
  src,
  alt,
  imgRef,
  imgStyles,
  placeholderStyles,
  spinnerHeight,
  spinnerWidth,
  spinnerColour,
}) => {
  const [loading, setLoading] = useState(true);

  return (
    <>
      {loading ? (
        <Placeholder $customStyles={placeholderStyles}>
          <ScaleLoader
            color={spinnerColour}
            height={spinnerHeight}
            width={spinnerWidth}
          />
        </Placeholder>
      ) : null}
      <Image
        src={src}
        alt={alt}
        ref={imgRef}
        onLoad={() => setLoading(false)}
        $customStyles={imgStyles}
        $loading={loading}
      />
    </>
  );
};

ImageLoader.propTypes = {
  src: PropTypes.string.isRequired,
  alt: PropTypes.string.isRequired,
  imgRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.instanceOf(Element) }),
  ]),
  imgStyles: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  placeholderStyles: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  spinnerHeight: PropTypes.number,
  spinnerWidth: PropTypes.number,
  spinnerColour: PropTypes.string,
};

export default ImageLoader;
