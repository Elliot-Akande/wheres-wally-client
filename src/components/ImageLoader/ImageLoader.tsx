import { forwardRef, useState } from "react";
import { ScaleLoader } from "react-spinners";
import { Interpolation } from "styled-components";
import { Image, Placeholder } from "./styles";

interface Props {
  src: string;
  alt: string;
  imgStyles?: Interpolation<React.CSSProperties>;
  placeholderStyles?: Interpolation<React.CSSProperties>;
  spinnerHeight?: number;
  spinnerWidth?: number;
  spinnerColour?: string;
}

type Ref = HTMLImageElement;

const ImageLoader = forwardRef<Ref, Props>(
  (
    {
      src,
      alt,
      imgStyles,
      placeholderStyles,
      spinnerHeight,
      spinnerWidth,
      spinnerColour,
    },
    ref
  ) => {
    const [loading, setLoading] = useState(true);

    return (
      <>
        {loading ? (
          <Placeholder $customStyles={placeholderStyles}>
            <ScaleLoader
              color={spinnerColour}
              height={spinnerHeight}
              width={spinnerWidth}
            />
          </Placeholder>
        ) : null}
        <Image
          src={src}
          alt={alt}
          ref={ref}
          onLoad={() => setLoading(false)}
          $customStyles={imgStyles}
          $loading={loading}
        />
      </>
    );
  }
);

export default ImageLoader;
