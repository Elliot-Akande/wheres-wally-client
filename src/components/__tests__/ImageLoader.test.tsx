import { fireEvent, render, screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import ImageLoader from "../ImageLoader/ImageLoader";

vi.mock("react-spinners", () => ({
  ScaleLoader: () => <div data-testid="spinner">Spinner</div>,
}));

describe("ImageLoader", () => {
  it("renders the spinner when loading", () => {
    const { container } = render(<ImageLoader src="#" alt="test" />);
    const img = screen.getByRole("img", { name: "test" });

    expect(img.complete).toBe(false);
    expect(container).toMatchSnapshot();
  });

  it("renders image and removes spinner after image loads", async () => {
    const { container } = render(<ImageLoader src="#" alt="test" />);

    const img = screen.getByRole("img", { name: "test" });
    fireEvent.load(img);

    const spinner = screen.queryByTestId("spinner");
    expect(spinner).toBeNull();
    expect(container).toMatchSnapshot();
  });
});
