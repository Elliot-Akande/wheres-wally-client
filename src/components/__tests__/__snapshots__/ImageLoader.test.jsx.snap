// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImageLoader > renders image and removes spinner after image loads 1`] = `
<div>
  <img
    alt="test"
    class="sc-guDLey cIQHQZ"
    src="#"
  />
</div>
`;

exports[`ImageLoader > renders the spinner when loading 1`] = `
<div>
  <div
    class="sc-beySPh jAEnzm"
  >
    <div
      data-testid="spinner"
    >
      Spinner
    </div>
  </div>
  <img
    alt="test"
    class="sc-guDLey hebdIY"
    src="#"
  />
</div>
`;
